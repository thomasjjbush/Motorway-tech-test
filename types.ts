import { FormEvent, ReactNode } from 'react';
import HtmlWebPackPlugin from 'html-webpack-plugin';
import { Json } from 'enzyme-to-json';

export interface AspectRatioProps {
    children: ReactNode;
    ratio: AspectRatios;
}

export enum AspectRatios {
    '1:1' = 100,
    '16:9' = 56.25,
    '4:3' = 75,
}

export interface ButtonProps {
    children: string;
    disabled?: boolean;
    onClick?: () => void;
    type?: 'button' | 'submit' | 'reset';
}

export interface FormField {
    label: string;
    max?: number;
    min?: number;
    placeholder?: string;
    type?: string;
    valid?: boolean;
    validation?: RegExp;
    value?: number | string;
}

export interface FormProps {
    ariaLabel: string;
    fields: {
        [field: string]: FormField;
    };
    onSubmit: (e: FormEvent) => void;
}

export interface Image {
    id: string;
    created_at: string;
    updated_at: string;
    color: string;
    description: string;
    alt_description: string;
    categories: string[];
    likes: number;
    user: {
        id: string;
        updated_at: string;
        username: string;
        name: string;
        first_name: string;
        last_name: string;
        bio: string;
        location: string;
        profile_image: string;
        total_collections: number;
        total_likes: number;
        total_photos: number;
    };
    url: string;
}

export interface ImageProps {
    alt: string;
    round?: boolean;
    src: string;
    width?: string;
}

export interface ImagesProps {
    columns: number;
}

export interface SerializerMap extends Json {
    node: {
        type: {
            componentStyle?: {
                baseStyle?: {
                    rules: (string | ((props: StyledProps) => string))[];
                };
                rules: (string | ((props: StyledProps) => string))[];
            };
        };
    };
}

export type StyledProps<P = unknown> = { theme: Theme } & P;

export interface Theme {
    colors: {
        blue: string;
        red: string;
        yellow: string;
    };
}

export interface WebpackConfig {
    devServer: { historyApiFallback: boolean; open: boolean };
    entry: string[];
    mode: string;
    module: {
        rules: WebpackRule[];
    };
    output: { path?: string; publicPath: string };
    plugins: any[];
    resolve: { extensions: string[] };
}

interface WebpackRule {
    exclude?: (RegExp | string)[];
    test: RegExp;
    use: {
        loader: string;
        options?: { plugins: string[] };
    };
}
