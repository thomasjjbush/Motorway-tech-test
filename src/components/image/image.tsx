import React, { FC, ReactElement } from 'react';
import styled from 'styled-components';
import { ImageProps as Props } from '../../../types';

const Img = styled.img<{ round: boolean }>`
    display: block;
    ${({ round }) => round && 'border-radius: 50%;'}
`;

export const Image: FC<Props> = ({ alt, round, src, width }: Props): ReactElement => {
    return (
        <picture>
            <source srcSet={`${src}.webp`} type="image/webp" />
            <Img alt={alt} data-test-id="img" round={round} src={`${src}.jpg`} width={width ?? '100%'} />
        </picture>
    );
};
