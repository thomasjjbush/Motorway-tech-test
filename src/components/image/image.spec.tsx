import React from 'react';
import { shallow } from 'enzyme';
import { Image } from './image';
import { ImageProps as Props } from '../../../types';

describe('Image', () => {
    const props: Props = {
        alt: 'alt',
        src: 'path/to/img',
    };

    it('should match snapshot', () => {
        expect(shallow(<Image {...props} />)).toMatchSnapshot();
    });

    it('should add add correct extensons', () => {
        const wrapper = shallow(<Image {...props} />);
        expect(wrapper.find({ 'data-test-id': 'img' }).prop('src')).toBe('path/to/img.jpg');
        expect(wrapper.find('source').prop('srcSet')).toBe('path/to/img.webp');
    });

    it('should add supplied width correctly', () => {
        expect(
            shallow(<Image {...props} width="helloWorld" />)
                .find({ 'data-test-id': 'img' })
                .prop('width'),
        ).toBe('helloWorld');
    });

    it('should apply border radius if round prop is provided and is true', () => {
        expect(shallow(<Image {...props} round />)).toMatchSnapshot();
    });
});
