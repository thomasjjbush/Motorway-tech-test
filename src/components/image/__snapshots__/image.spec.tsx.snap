// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should apply border radius if round prop is provided and is true 1`] = `
<picture>
  <source
    srcSet="path/to/img.webp"
    type="image/webp"
  />
  <styled.img
    alt="alt"
    data-test-id="img"
    round={true}
    src="path/to/img.jpg"
    styles="
    display: block;
    border-radius: 50%;
"
    width="100%"
  />
</picture>
`;

exports[`Image should match snapshot 1`] = `
<picture>
  <source
    srcSet="path/to/img.webp"
    type="image/webp"
  />
  <styled.img
    alt="alt"
    data-test-id="img"
    src="path/to/img.jpg"
    styles="
    display: block;
    
"
    width="100%"
  />
</picture>
`;
